File: Smartphone_App_Pipeline/lkr-data/lib/wearable/gaitPhase.ts
import {
  EARLY_STANCE,
  LATE_STANCE,
  MIDDLE_STANCE,
  SWING,
} from '@/lib/wearable/const';
import type { SensorData } from '@/lib/wearable/sensorTypes';

function gyroMagNorm(sensorData: SensorData): number {
  const { GyroX, GyroY, GyroZ } = sensorData;
  return Math.sqrt(GyroX * GyroX + GyroY * GyroY + GyroZ * GyroZ);
}

/** Port of Bluetooth/gaitphase.py */
export class GaitPhase {
  lastStanceTime = 0.6;
  readonly middlestanceItersThreshold: number;
  readonly latestanceItersThreshold: number;
  readonly gyromagThresholdHeelstrike = 45;
  readonly gyromagThresholdToeoff = 45;
  readonly heelstrikeItersThreshold: number;
  readonly datarate: number;

  gaitphase = LATE_STANCE;
  gaitphaseOld = LATE_STANCE;
  stepCount = 0;
  itersConsecutiveBelowGyroMagThresh = 0;
  itersStance = 0;

  inFeedbackWindow = false;

  readonly fpaBuffer: number[] = [];
  fpaThisFrame = 0;
  fpaThisStep = 0;

  constructor(datarate = 50) {
    this.datarate = datarate;
    this.lastStanceTime = 0.6;
    this.middlestanceItersThreshold = this.lastStanceTime * 0.25 * datarate;
    this.latestanceItersThreshold = this.lastStanceTime * 0.5 * datarate;
    this.heelstrikeItersThreshold = 0.1 * datarate;
  }

  updateGaitphase(sensorData: SensorData): void {
    const gyroMag = gyroMagNorm(sensorData);

    if (this.gaitphase === SWING) {
      this.gaitphaseOld = SWING;
      if (gyroMag < this.gyromagThresholdHeelstrike) {
        this.itersConsecutiveBelowGyroMagThresh += 1;
        if (this.itersConsecutiveBelowGyroMagThresh > this.heelstrikeItersThreshold) {
          this.itersConsecutiveBelowGyroMagThresh = 0;
          this.itersStance = 0;
          this.stepCount += 1;
          this.gaitphase = EARLY_STANCE;
        }
      } else {
        this.itersConsecutiveBelowGyroMagThresh = 0;
      }
    } else if (this.gaitphase === EARLY_STANCE) {
      this.gaitphaseOld = EARLY_STANCE;
      this.itersStance += 1;
      if (this.itersStance > this.middlestanceItersThreshold) {
        this.gaitphase = MIDDLE_STANCE;
      }
    } else if (this.gaitphase === MIDDLE_STANCE) {
      this.gaitphaseOld = MIDDLE_STANCE;
      this.itersStance += 1;
      if (this.itersStance > this.latestanceItersThreshold) {
        this.gaitphase = LATE_STANCE;
      }
    } else if (this.gaitphase === LATE_STANCE) {
      this.gaitphaseOld = LATE_STANCE;
      this.itersStance += 1;
      if (gyroMag > this.gyromagThresholdToeoff) {
        this.lastStanceTime = this.itersStance / this.datarate;
        if (this.lastStanceTime > 2) {
          this.lastStanceTime = 2;
        } else if (this.lastStanceTime < 0.4) {
          this.lastStanceTime = 0.4;
        }
        this.gaitphase = SWING;
      }
    }

    this.inFeedbackWindow =
      this.gaitphaseOld === MIDDLE_STANCE && this.gaitphase === LATE_STANCE;
  }
}

File: Laptop_Pipeline/algorithms/sage_motion/gaitphase.py
#TAKEN FROM SAGEMOTION: https://github.com/SageMotionApps/004_Walking_Foot_Progression_Angle.git

import numpy as np
from algorithms.base import EARLY_STANCE, MIDDLE_STANCE, LATE_STANCE, SWING


class GaitPhase:
    def __init__(self, datarate=50):
        self.last_stance_time = 0.6
        self.MIDDLESTANCE_ITERS_THRESHOLD = self.last_stance_time * 0.25 * datarate
        self.LATESTANCE_ITERS_THRESHOLD = self.last_stance_time * 0.5 * datarate
        self.GYROMAG_THRESHOLD_HEELSTRIKE = 45
        self.GYROMAG_THRESHOLD_TOEOFF = 45
        self.HEELSTRIKE_ITERS_THRESHOLD = 0.1 * datarate
        self.DATARATE = datarate

        self.gaitphase = LATE_STANCE
        self.gaitphase_old = LATE_STANCE
        self.step_count = 0
        self.iters_consecutive_below_gyroMag_thresh = 0
        self.iters_stance = 0

        self.in_feedback_window = False

        self.FPA_buffer = []
        self.FPA_this_frame = 0
        self.FPA_this_step = 0

    def update_gaitphase(self, sensor_data):
        gyroMag = np.linalg.norm(
            [sensor_data["GyroX"], sensor_data["GyroY"], sensor_data["GyroZ"]],
            ord=2,
        )
        if self.gaitphase == SWING:
            self.gaitphase_old = SWING
            if gyroMag < self.GYROMAG_THRESHOLD_HEELSTRIKE:
                self.iters_consecutive_below_gyroMag_thresh += 1
                if (
                    self.iters_consecutive_below_gyroMag_thresh
                    > self.HEELSTRIKE_ITERS_THRESHOLD
                ):
                    self.iters_consecutive_below_gyroMag_thresh = 0
                    self.iters_stance = 0
                    self.step_count += 1
                    self.gaitphase = EARLY_STANCE
            else:
                self.iters_consecutive_below_gyroMag_thresh = 0
        elif self.gaitphase == EARLY_STANCE:
            self.gaitphase_old = EARLY_STANCE
            self.iters_stance += 1
            if self.iters_stance > self.MIDDLESTANCE_ITERS_THRESHOLD:
                self.gaitphase = MIDDLE_STANCE
        elif self.gaitphase == MIDDLE_STANCE:
            self.gaitphase_old = MIDDLE_STANCE
            self.iters_stance += 1
            if self.iters_stance > self.LATESTANCE_ITERS_THRESHOLD:
                self.gaitphase = LATE_STANCE
        elif self.gaitphase == LATE_STANCE:
            self.gaitphase_old = LATE_STANCE
            self.iters_stance += 1
            if gyroMag > self.GYROMAG_THRESHOLD_TOEOFF:
                self.last_stance_time = self.iters_stance / self.DATARATE
                if self.last_stance_time > 2:
                    self.last_stance_time = 2
                elif self.last_stance_time < 0.4:
                    self.last_stance_time = 0.4
                self.gaitphase = SWING

        self.in_feedback_window = (
            self.gaitphase_old == MIDDLE_STANCE and self.gaitphase == LATE_STANCE
        )
